namecache/namecache.go

```diff
@@ -0,0 +1,254 @@
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	cancel, ok := nc.follows[name]
+	if !ok {
+		return fmt.Errorf("Unknown name %s", name)
+	}
+
+	cancel()
+	delete(nc.follows, name)
+	return nil
+}
+
+// ListFollows returns a list of names currently being followed
+func (nc *nameCache) ListFollows() []string {
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	follows := make([]string, len(nc.follows))[:0]
+	for name, _ := range nc.follows {
+		follows = append(follows, name)
+	}
+
+	return follows
+}
+
+func (nc *nameCache) followName(ctx context.Context, name string, pinit bool, followInterval time.Duration) {
+	// if cid != nil, we have created a new pin that is updated on changes and
+	// unpinned on cancel
+	cid, err := nc.resolveAndPin(ctx, name, pinit)
+	if err != nil {
+		log.Errorf("Error following %s: %s", name, err.Error())
+	}
+
+	ticker := time.NewTicker(followInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			if cid != nil {
+				cid, err = nc.resolveAndUpdate(ctx, name, cid)
+			} else {
+				cid, err = nc.resolveAndPin(ctx, name, pinit)
+			}
+
+			if err != nil {
+				log.Errorf("Error following %s: %s", name, err.Error())
+			}
+
+		case <-ctx.Done():
+			if cid != nil {
+				err = nc.unpin(cid)
+				if err != nil {
+					log.Errorf("Error unpinning %s: %s", name, err.Error())
+				}
+			}
+			return
+		}
+	}
+}
+
+func (nc *nameCache) resolveAndPin(ctx context.Context, name string, pinit bool) (*cid.Cid, error) {
+	ptr, err := nc.resolve(ctx, name)
+	if err != nil {
+		return nil, err
+	}
+
+	if !pinit {
+		return nil, nil
+	}
+
+	cid, err := pathToCid(ptr)
+	if err != nil {
+		return nil, err
+	}
+
+	defer nc.bstore.PinLock().Unlock()
+
+	_, pinned, err := nc.pinning.IsPinned(cid)
+	if pinned || err != nil {
+		return nil, err
+	}
+
+	n, err := nc.pathToNode(ctx, ptr)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Debugf("pinning %s", cid.String())
+
+	err = nc.pinning.Pin(ctx, n, true)
+	if err != nil {
+		return nil, err
+	}
+
+	err = nc.pinning.Flush()
+
+	return cid, err
+}
+
+func (nc *nameCache) resolveAndUpdate(ctx context.Context, name string, cid *cid.Cid) (*cid.Cid, error) {
+
+	ptr, err := nc.resolve(ctx, name)
+	if err != nil {
+		return nil, err
+	}
+
+	ncid, err := pathToCid(ptr)
+	if err != nil {
+		return nil, err
+	}
+
+	if ncid.Equals(cid) {
+		return cid, nil
+	}
+
+	defer nc.bstore.PinLock().Unlock()
+
+	log.Debugf("Updating pin %s -> %s", cid.String(), ncid.String())
+
+	err = nc.pinning.Update(ctx, cid, ncid, true)
+	if err != nil {
+		return cid, err
+	}
+
+	err = nc.pinning.Flush()
+
+	return ncid, err
+}
+
+func (nc *nameCache) unpin(cid *cid.Cid) error {
+	defer nc.bstore.PinLock().Unlock()
+
+	err := nc.pinning.Unpin(nc.ctx, cid, true)
+	if err != nil {
+		return err
+	}
+
+	return nc.pinning.Flush()
+}
+
+func (nc *nameCache) resolve(ctx context.Context, name string) (path.Path, error) {
+	log.Debugf("resolving %s", name)
+
+	if !strings.HasPrefix(name, "/ipns/") {
```

Comment:
Can we do this at a higher layer?

Reply:
Probably, but does it matter?

Reply:
Sanitize data as soon as it enters the program. I also advocate moving this into the command's `Run` function above.

Reply:
The `Run` function is perhaps inappropriate, it doesn't cover direct uses of the API. 
If we are to enforce the `/ipns/` prefix then it should be done at the `Namecache` interface.

Still, it  will have to be done in 3 places instead of 1.

Reply:
I moved it to the `Namecache` interface methods (`Add` and `Cancel`).